src/components/LinkChooser/LinkChooser.tsx

```diff
@@ -0,0 +1,218 @@
+import { Popover } from "./Popover";
+import { SearchInput } from "./SearchInput";
+import { SearchResultsList } from "./SearchResultSection";
+import { SectionActionMenu } from "./SectionActionMenu";
+import { linkChooserMachine, LinkChooserState } from "./state/machine";
+import { LinkChooserProps } from "./types";
+import * as SearchRepository from "./repositories";
+import { createCustomLink, isLoaded, queryMatchesSelection } from "./utils/helpers";
+
+export enum IconLabel {
+    Document = "DOCUMENT",
+    Library = "LIBRARY",
+    Link = "LINK",
+    External = "EXTERNAL",
+    Template = "TEMPLATE",
+}
+
+export const ICON_OPTIONS: Record<IconLabel | string, ReactElement> = {
+    [IconLabel.Document]: <IconDocument />,
+    [IconLabel.Library]: <IconDocumentLibrary />,
+    [IconLabel.Link]: <IconLink />,
+    [IconLabel.External]: <IconExternalLink />,
+    [IconLabel.Template]: <IconTemplate />,
+};
+
+export const DEFAULT_ICON = IconLabel.Link;
+export const CUSTOM_LINK_ID = "custom-link";
+export const MAX_STORED_ITEMS = 5;
+export const QUERIES_STORAGE_KEY = "queries";
+
+export const LinkChooser: FC<LinkChooserProps> = ({
+    getGlobalByQuery = SearchRepository.getGlobalByQuery,
+    getTemplatesByQuery = SearchRepository.getTemplatesByQuery,
+    openPreview = window.open,
+    copyToClipboard = navigator.clipboard,
+    openInNewTab,
+    ariaLabel = "Menu",
+    label,
+    placeholder,
+    onOpenInNewTabChange,
+    onLinkChange,
+}) => {
+    const [{ context, matches, value }, send, service] = useMachine(
+        linkChooserMachine.withContext({
+            searchResults: [],
+            selectedResult: null,
+            query: "",
+            interruptedFetch: false,
+            copyToClipboard,
+            openPreview,
+            getGlobalByQuery,
+            getTemplatesByQuery,
+            onLinkChange,
+        }),
+    );
+
+    const handleClearClick = useCallback(() => {
+        state.setInputValue("");
+        state.setSelectedKey("");
+        send("CLEARING", { data: { query: "" } });
+    }, []);
+
+    const handleSelectionChange = (key: Key) => {
+        const foundItem = context.searchResults.find((item) => item.id === key);
+        if (foundItem) send("SET_SELECTED_SEARCH_RESULT", { data: { selectedResult: foundItem } });
+    };
+
+    const handleInputChange = (value: string) => {
+        send("TYPING", { data: { query: value } });
+    };
+
+    const handleDropdownOpen = () => {
+        send("OPEN_DROPDOWN");
+    };
+
+    const handleDropdownClose = () => {
+        const selectedResult = context.query
+            ? queryMatchesSelection(context.selectedResult, context.query)
+                ? context.selectedResult
+                : createCustomLink(state.inputValue)
+            : null;
+        send("CLOSE_DROPDOWN", { data: { selectedResult } });
+        state.setSelectedKey(selectedResult?.id || "");
```

Comment:
the reason why the abovementioned methods appear before the state is defined is because i wanted to organize all of the event handlers together in one place and not have them scattered all over the place inside the component.
but if that will be an improvement, i will move them after the state instantiation